File: go/api/Pb.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Pb 对应 Pb.py。
type Pb struct {
	*session.ApiGroup
}

// NewPb 创建 Pb API 分组。
func NewPb(s *session.Session) *Pb {
	return &Pb{ApiGroup: session.NewApiGroup(s)}
}

// GetPbMatch 在电话簿中查找号码。对应 pb/pb-match (post_get)。
func (p *Pb) GetPbMatch(phoneNumber string) (map[string]interface{}, error) {
	return p.S.PostGet("pb/pb-match", map[string]interface{}{
		"Phone": phoneNumber,
	}, false, "api", false, false)
}

// GetPbList 获取电话簿条目列表。对应 pb/pb-list (post_get)。
func (p *Pb) GetPbList(
	page int,
	keyWord string,
	groupId int,
	readCount int,
	saveType int,
	sortType int,
	ascending int,
) (map[string]interface{}, error) {
	return p.S.PostGet("pb/pb-list", session.O(
		"GroupID", groupId,
		"PageIndex", page,
		"ReadCount", readCount,
		"SaveType", saveType,
		"SortType", sortType,
		"Ascending", ascending,
		"KeyWord", keyWord,
	), false, "api", false, false)
}

// PbCount 获取电话簿条目数量。对应 pb/pb-count (post_get)。
func (p *Pb) PbCount() (map[string]interface{}, error) {
	return p.S.PostGet("pb/pb-count", nil, false, "api", false, false)
}

// GroupCount 获取电话簿分组数量。对应 pb/group-count (post_get)。
func (p *Pb) GroupCount() (map[string]interface{}, error) {
	return p.S.PostGet("pb/group-count", nil, false, "api", false, false)
}

// pbField 电话簿联系人字段节点，对应 Python 的 Node("Field")。
// 注意 Python 中每个 Node 实例都是独立 key，因此 5 个 Field 全部序列化。
func pbField(name, value string) *session.ODict {
	return session.O("Name", name, "Value", value)
}

// PbNew 添加新的电话簿条目。对应 pb/pb-new (post_set)。
//
// 注意：Python 版使用 5 个独立的 Node("Field") key，Go 版用 []*ODict
// 输出重复的 <Field> 节点，行为一致。
func (p *Pb) PbNew(
	groupId int,
	saveType int,
	name string,
	mobilePhone string,
	homePhone string,
	workPhone string,
	workEmail string,
) (interface{}, error) {
	return p.S.PostSet("pb/pb-new", session.O(
		"GroupID", groupId,
		"SaveType", saveType,
		"Field", []*session.ODict{
			pbField("FormattedName", name),
			pbField("MobilePhone", mobilePhone),
			pbField("HomePhone", homePhone),
			pbField("WorkPhone", workPhone),
			pbField("WorkEmail", workEmail),
		},
	), false, "api", false, false)
}

// PbDelete 按 ID 删除电话簿条目。对应 pb/pb-delete (post_set)。
func (p *Pb) PbDelete(pbIndex int) (interface{}, error) {
	return p.S.PostSet("pb/pb-delete", map[string]interface{}{
		"Index": pbIndex,
	}, false, "api", false, false)
}

// GroupDelete 按 ID 删除电话簿分组。对应 pb/group-delete (post_set)。
func (p *Pb) GroupDelete(groupId int) (interface{}, error) {
	return p.S.PostSet("pb/group-delete", map[string]interface{}{
		"GroupID": groupId,
	}, false, "api", false, false)
}

// GroupList 获取电话簿分组列表。对应 pb/group-list (post_get)。
func (p *Pb) GroupList(
	page int,
	readCount int,
	sortType int,
	ascending int,
) (map[string]interface{}, error) {
	return p.S.PostGet("pb/group-list", session.O(
		"PageIndex", page,
		"ReadCount", readCount,
		"SortType", sortType,
		"Ascending", ascending,
	), false, "api", false, false)
}

// GroupNew 按名称创建电话簿分组。对应 pb/group-new (post_set)。
func (p *Pb) GroupNew(nameStr string) (interface{}, error) {
	return p.S.PostSet("pb/group-new", map[string]interface{}{
		"GroupName": nameStr,
	}, false, "api", false, false)
}

File: src/api/pb.rs
//! Pb API group (`api/Pb.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::{XmlMap, XmlValue};

use super::{post_get_value, post_set};

/// Pb API group.
pub struct Pb<'a> {
    conn: &'a Connection,
}

impl<'a> Pb<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Pb { conn }
    }

    /// Find a number in the phone book.
    pub fn get_pb_match(&self, phone_number: &str) -> Result<Value> {
        let body: XmlMap = map_of([("Phone", phone_number.to_string())]);
        post_get_value(self.conn, "api/pb/pb-match", &body)
    }

    /// Get the list of phone book entries.
    #[allow(clippy::too_many_arguments)]
    pub fn get_pb_list(
        &self,
        page: i64,
        key_word: &str,
        group_id: i64,
        read_count: i64,
        save_type: i64,
        sort_type: i64,
        ascending: i64,
    ) -> Result<Value> {
        let body: XmlMap = map_of([
            ("GroupID", group_id.to_string()),
            ("PageIndex", page.to_string()),
            ("ReadCount", read_count.to_string()),
            ("SaveType", save_type.to_string()),
            ("SortType", sort_type.to_string()),
            ("Ascending", ascending.to_string()),
            ("KeyWord", key_word.to_string()),
        ]);
        post_get_value(self.conn, "api/pb/pb-list", &body)
    }

    /// Count of phone book entries (reverse engineered).
    pub fn pb_count(&self) -> Result<Value> {
        let body = XmlMap::new();
        post_get_value(self.conn, "api/pb/pb-count", &body)
    }

    /// Count of phone book groups (reverse engineered).
    pub fn group_count(&self) -> Result<Value> {
        let body = XmlMap::new();
        post_get_value(self.conn, "api/pb/group-count", &body)
    }

    /// Add a new entry to the global phone book.
    #[allow(clippy::too_many_arguments)]
    pub fn pb_new(
        &self,
        group_id: i64,
        save_type: i64,
        name: &str,
        mobile_phone: &str,
        home_phone: &str,
        work_phone: &str,
        work_email: &str,
    ) -> Result<String> {
        let mut field_maps = Vec::new();
        for (field_name, value) in [
            ("FormattedName", name),
            ("MobilePhone", mobile_phone),
            ("HomePhone", home_phone),
            ("WorkPhone", work_phone),
            ("WorkEmail", work_email),
        ] {
            let mut f = XmlMap::new();
            f.insert("Name".into(), XmlValue::Text(field_name.to_string()));
            f.insert("Value".into(), XmlValue::Text(value.to_string()));
            field_maps.push(XmlValue::Map(f));
        }

        let mut body = XmlMap::new();
        body.insert("GroupID".into(), XmlValue::Text(group_id.to_string()));
        body.insert("SaveType".into(), XmlValue::Text(save_type.to_string()));
        body.insert("Field".into(), XmlValue::List(field_maps));
        post_set(self.conn, "api/pb/pb-new", &body)
    }

    /// Delete a phone book entry by its index.
    pub fn pb_delete(&self, pb_index: i64) -> Result<String> {
        let body: XmlMap = map_of([("Index", pb_index.to_string())]);
        post_set(self.conn, "api/pb/pb-delete", &body)
    }

    /// Delete a phone book group by its ID.
    pub fn group_delete(&self, group_id: i64) -> Result<String> {
        let body: XmlMap = map_of([("GroupID", group_id.to_string())]);
        post_set(self.conn, "api/pb/group-delete", &body)
    }

    /// Get the list of phone book groups.
    pub fn group_list(
        &self,
        page: i64,
        read_count: i64,
        sort_type: i64,
        ascending: i64,
    ) -> Result<Value> {
        let body: XmlMap = map_of([
            ("PageIndex", page.to_string()),
            ("ReadCount", read_count.to_string()),
            ("SortType", sort_type.to_string()),
            ("Ascending", ascending.to_string()),
        ]);
        post_get_value(self.conn, "api/pb/group-list", &body)
    }

    /// Create a new phone book group by name.
    pub fn group_new(&self, name_str: &str) -> Result<String> {
        let body: XmlMap = map_of([("GroupName", name_str.to_string())]);
        post_set(self.conn, "api/pb/group-new", &body)
    }
}
